xvec/index.py

```diff
@@ -0,0 +1,193 @@
+    :class:`xarray.indexes.PandasIndex`. In addition, it allows spatial
+    filtering based on geometries (powered by :class:`shapely.STRtree`).
+
+    Parameters
+    ----------
+    index : :class:`xarray.indexes.PandasIndex`
+        An Xarray (pandas) index built from an array-like of
+        :class:`shapely.Geometry` objects.
+    crs : object
+        The coordinate reference system. Any value accepted by
+        :meth:`pyproj.crs.CRS.from_user_input`.
+
+    """
+
+    _index: PandasIndex
+    _sindex: shapely.STRtree | None
+    _crs: CRS
+
+    def __init__(self, index: PandasIndex, crs: CRS):
+        if not np.all(shapely.is_geometry(index.index)):
+            raise ValueError("array must contain shapely.Geometry objects")
+
+        self._crs = CRS.from_user_input(crs)
+        self._index = index
+        self._sindex = None
+
+    @property
+    def crs(self) -> CRS:
+        """Returns the coordinate reference system of the index as a
+        :class:`pyproj.crs.CRS` object.
+        """
+        return self._crs
+
+    @property
+    def sindex(self) -> shapely.STRtree:
+        """Returns the spatial index, i.e., a :class:`shapely.STRtree` object.
+
+        It may build the index before returning it.
+        """
+        if self._sindex is None:
+            self._sindex = shapely.STRtree(self._index.index)
+        return self._sindex
+
+    @classmethod
+    def from_variables(
+        cls,
+        variables: Mapping[Any, Variable],
+        *,
+        options: Mapping[str, Any],
+    ):
+        # TODO: try getting CRS from coordinate attrs or GeometryArray
+        if "crs" not in options:
+            raise ValueError("a CRS must be provided")
+
+        index = PandasIndex.from_variables(variables, options={})
+        return cls(index, crs=options["crs"])
+
+    @classmethod
+    def concat(
+        cls,
+        indexes: Sequence[GeoVectorIndex],
+        dim: Hashable,
+        positions: Iterable[Iterable[int]] | None = None,
+    ) -> GeoVectorIndex:
+        crss = [idx.crs for idx in indexes]
+
+        if any([s != crss[0] for s in crss]):
+            raise ValueError("conflicting CRS for coordinates to concat")
+
+        indexes_ = [idx._index for idx in indexes]
+        index = PandasIndex.concat(indexes_, dim, positions)
+        return cls(index, crss[0])
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Hashable, Variable]:
+        return self._index.create_variables(variables)
+
+    def to_pandas_index(self) -> pd.Index:
+        return self._index.index
+
+    def isel(self, indexers: Mapping[Any, Any]):
+        index = self._index.isel(indexers)
+
+        if index is not None:
+            return type(self)(index, self.crs)
+        else:
+            return None
+
+    def _sel_sindex(self, labels, method, tolerance):
+        # only one coordinate supported
+        assert len(labels) == 1
```

Comment:
Can we turn this into a ValueError?

Reply:
It should always be equal to 1 here (Xarray already filter the labels passed to each index). Let's keep the `assert` statement for debugging and for clarity?